### DISCUSSION TITLE
Costmaps/OccupancyGrid visualization

### DISCUSSION THREAD

**Discussion by charlielito** [foxglove/community#67] [Q&A]
Hi everyone!  
Costmaps play an important role in some autonomous navigation stacks. Being able to visualize them properly is very important to debug those systems. Currently, I am trying to use Foxglove for that purpose to visualize `nav_msgs/msg/OccupancyGrid` costmaps. 

Nonetheless, the visualization is not as detailed as it is in Rviz. An example side-by-side of both visualizations using ROS2 galactic:

![2022-04-06_20-32](https://user-images.githubusercontent.com/8033598/162112036-0af432de-512f-4192-84a1-4f5501f7e81c.png)
 
As you can see there, the different values of the costmap are not that clear in the Foxglove one.

Are any plans to support more rviz-like occupancygrid visualizations?

**Comment by defunctzombie** [comment]
For the `rviz` visualization - are you able to configure different colors for different ranges of values? Is that how you end up with the colorings you are showing? My rough understanding of OccupancyGrid was that the data was a floating point value typically treated as a gradient/grayscale.

/cc @jtbandes who's doing some work with Grids in foxglove and might be interested in hearing more use cases like this one.

**Reply by charlielito** [reply]
Yes. Actually, there are 3 `color schemes` that the user can choose:
* `Map`
![2022-04-07_11-36_1](https://user-images.githubusercontent.com/8033598/162248032-b749aae0-f6a8-4278-a462-3003045f0a7d.png)


* `Costmap`
![2022-04-07_11-36](https://user-images.githubusercontent.com/8033598/162248060-780fe750-3020-484f-bcae-f4e1b13be401.png)

* `Raw`
![2022-04-07_11-37](https://user-images.githubusercontent.com/8033598/162248085-51dc3116-778a-4752-9dab-9a956424d878.png)

It seems Foxglove is using the `map` color scheme.

OccupancyGrids are of type `uint8` and the color scheme define how they will look. Look how rviz implemented that color scheme: https://github.com/ros2/rviz/blob/c668ff1a7122dd806bdb6ef09593e1f451f8bd97/rviz_default_plugins/src/rviz_default_plugins/displays/map/palette_builder.cpp#L101

It would be nice if Foxglove could provide the user with a similar user interface to choose the color scheme

**Reply by jtbandes** [reply]
Thanks for the feedback, I agree it would be great to have this level of customizability (and even more).

We have just expanded our supported types with a custom `foxglove.Grid` message: https://github.com/foxglove/studio/pull/3136

For this custom grid and regular OccupancyGrids, we would like to expand the available settings as you describe. Practically, this will be dependent on two major work streams we're currently working on: [a new 3D panel](https://github.com/foxglove/studio/issues/3143), and [a new settings sidebar](https://github.com/foxglove/studio/pull/3150). As these both evolve it will become easier for us to add new features and settings to the different visualization types.

**Reply by charlielito** [reply]
Nice! Excited to see those features!
